python/ray/data/_internal/batcher.py

```diff
@@ -275,46 +267,63 @@ def has_batch(self) -> bool:
         buffer_size = self._buffer_size()
 
         if not self._done_adding:
-            # If still adding blocks, ensure that removing a batch wouldn't cause the
-            # shuffle buffer to dip beneath its configured minimum size.
-            return buffer_size - self._batch_size >= self._buffer_min_size
+            # Delay pulling of batches until the buffer is large enough in order to
+            # amortize compaction overhead.
+            return (
+                buffer_size - self._batch_size
+                >= self._buffer_min_size * SHUFFLE_BUFFER_COMPACTION_RATIO
```

Comment:
what does `buffer_size - self._batch_size` mean? I think here we should compare `self._builder.num_rows()` against something.

Reply:
buffer size is both the rows in the builder and materialized buffer. We need to preserve the invariant that the shuffle buffer window is always at least the specified size after removing a batch, which explains this logic.